packages/components/src/input-control/reducer/reducer.ts

```diff
@@ -2,69 +2,58 @@
  * External dependencies
  */
 import { isEmpty } from 'lodash';
+// eslint-disable-next-line no-restricted-imports
+import type { SyntheticEvent } from 'react';
+
 /**
  * WordPress dependencies
  */
 import { useReducer } from '@wordpress/element';
 
-const initialStateReducer = ( state ) => state;
-
-const initialInputControlState = {
-	_event: {},
-	error: null,
-	initialValue: '',
-	isDirty: false,
-	isDragEnabled: false,
-	isDragging: false,
-	isPressEnterToChange: false,
-	value: '',
-};
-
-const actionTypes = {
-	CHANGE: 'CHANGE',
-	COMMIT: 'COMMIT',
-	DRAG_END: 'DRAG_END',
-	DRAG_START: 'DRAG_START',
-	DRAG: 'DRAG',
-	INVALIDATE: 'INVALIDATE',
-	PRESS_DOWN: 'PRESS_DOWN',
-	PRESS_ENTER: 'PRESS_ENTER',
-	PRESS_UP: 'PRESS_UP',
-	RESET: 'RESET',
-	UPDATE: 'UPDATE',
-};
-
-export const inputControlActionTypes = actionTypes;
+/**
+ * Internal dependencies
+ */
+import {
+	InputState,
+	StateReducer,
+	initialInputControlState,
+	initialStateReducer,
+} from './state';
+import * as actions from './actions';
 
 /**
  * Prepares initialState for the reducer.
  *
- * @param {Object} initialState The initial state.
- * @return {Object} Prepared initialState for the reducer
+ * @param  initialState The initial state.
+ * @return Prepared initialState for the reducer
  */
-function mergeInitialState( initialState = initialInputControlState ) {
+function mergeInitialState(
+	initialState: Partial< InputState > = initialInputControlState
+): InputState {
 	const { value } = initialState;
 
 	return {
 		...initialInputControlState,
 		...initialState,
 		initialValue: value,
-	};
+	} as InputState;
```

Comment:
Why does this need to be explicitly cast? TypeScript should be able to infer and combine the types from the other objects. If there's an error or something, there may be a bug there that the cast is hiding?

Reply:
It was because the `initialValue` couldn't be `undefined` and the `Partial<InputState>` was making it allowed to be undefined. But I think previously this was allowable anyway so I've made `value` and `initialValue` undefined.

Though won't that mess up the "controlled-ness" of the input? Should we only allow "empty" values but not undefined?

Reply:
I'm going to merge this PR and we can refine the types for this if we decide it's a problem. Currently I think we're not introducing anything that isn't already true for the intrinsic `input` element.